Row byte count is not available in GraphicsContextGL::computeImageSizeBytes

kkinnunen-apple opened the pull request:
#### 19a2252db08cadac7ce04a570c7d4be1db253245
<pre>
Row byte count is not available in GraphicsContextGL::computeImageSizeBytes
<a href="https://bugs.webkit.org/show_bug.cgi?id=257452">https://bugs.webkit.org/show_bug.cgi?id=257452</a>
rdar://109969560

Reviewed by Dan Glastonbury.

GraphicsContextGL::computeImageSizeInBytes would return the image size
but not the count of row bytes. The row bytes are needed for GPUP side
implementation where the GPUP is queried for tightly packed image
and then the image data is expanded by WP to the client buffer.

Rename to computeImageSize.

Simplify the computation by using Checked all the way to the end.
The idea of Checked is that the intermediate values do not need to be
checked.

Return a struct instead of numerous output variables. These were
error-prone, as they were written even though the function would return
error.

Remove the unused paddingInBytes, which was not used.
Add row bytes variables.

Add checked variant of roundUpToMultipleOfNonPowerOfTwo.
Add unchecked variant, for consistency.
Fix overflow in bmalloc::roundUpToMultipleOfNonPowerOfTwo and
untemplatize, as the template parameter could not be anything else than
size_t.

* Source/WTF/wtf/StdLibExtras.h:
(WTF::roundUpToMultipleOfNonPowerOfTwo):
* Source/WebCore/html/canvas/WebGL2RenderingContext.cpp:
(WebCore::WebGL2RenderingContext::readPixels):
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::readPixels):
(WebCore::WebGLRenderingContextBase::validateImageFormatAndType):
(WebCore::WebGLRenderingContextBase::validateTexFuncData):
* Source/WebCore/html/canvas/WebGLRenderingContextBase.h:
* Source/WebCore/platform/graphics/GraphicsContextGL.cpp:
(WebCore::GraphicsContextGL::computeImageSize):
(WebCore::GraphicsContextGL::packImageData):
(WebCore::GraphicsContextGL::extractPixelBuffer):
(WebCore::GraphicsContextGL::extractTextureData):
(WebCore::GraphicsContextGL::computeImageSizeInBytes): Deleted.
* Source/WebCore/platform/graphics/GraphicsContextGL.h:
* Source/bmalloc/bmalloc/Algorithm.h:
(bmalloc::roundUpToMultipleOfNonPowerOfTwo):
* Tools/TestWebKitAPI/Tests/WTF/StdLibExtras.cpp:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/264794@main">https://commits.webkit.org/264794@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5f359fb8ca6326ccbf899b0753a9a115ba10a19b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8617 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8907 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/9122 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/10273 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8634 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/8626 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10896 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8880 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11501 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8763 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9786 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/7725 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/10431 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/7086 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7880 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/15411 "1 flakes 115 failures") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/7413 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/8205 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/8028 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/11372 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/8265 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/8511 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6941 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/8791 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7781 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/2112 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2099 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11992 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/9020 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/8252 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/2232 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WTF/wtf/StdLibExtras.h:
```diff
@@ -205,6 +205,26 @@ template<size_t divisor, typename T> inline constexpr T* roundUpToMultipleOf(T*
+template<typename T, typename U>
+inline constexpr T roundUpToMultipleOfNonPowerOfTwo(U divisor, T x)
+{
+    auto reminder = x % divisor;
```
`remainder`

djg commented on Source/WTF/wtf/StdLibExtras.h:
```diff
@@ -205,6 +205,26 @@ template<size_t divisor, typename T> inline constexpr T* roundUpToMultipleOf(T*
 }
 
+template<typename T, typename U>
+inline constexpr T roundUpToMultipleOfNonPowerOfTwo(U divisor, T x)
```
Why does the divisor come first?
Should we use SFINAE to open enable this function for integral types?

kkinnunen-apple replied:
> Why does the divisor come first? 

Theres numerous similar functions, where the divisor does come first. Also the template parameters T, U are in that order in other functions.

> Should we use SFINAE to open enable this function for integral types?

Well should we ? :) If a thing is useful to mention as a review comment, it should be useful to spend the time and expand the comment with a sentence of a rationale.
I'm not sure what that means, SFINAE to open enable for integral types. Given that this function is rather similar to numerous other functions that do the exact same thing around it, and it works for this use-case, I'd say not in this patch? Maybe all the functions can be improved later on.

Note: the Checked and non-Checked cannot be supported by the same template, if that's what you mean. The checked variant currently has to check the overflow during the if and during the mod. Checked could be fixed for mod, but elaborate framework removing the if wouldn't let the same template be used.

djg commented on Source/WebCore/platform/graphics/GraphicsContextGL.cpp:
```diff
@@ -425,118 +425,62 @@ unsigned GraphicsContextGL::computeBytesPerGroup(GCGLenum format, GCGLenum type)
 }
 
-GCGLenum GraphicsContextGL::computeImageSizeInBytes(GCGLenum format, GCGLenum type, GCGLsizei width, GCGLsizei height, GCGLsizei depth, const PixelStoreParameters& params, unsigned* imageSizeInBytes, unsigned* paddingInBytes, unsigned* skipSizeInBytes)
+std::optional<GraphicsContextGL::PixelRectangleSizes> GraphicsContextGL::computeImageSize(GCGLenum format, GCGLenum type, IntSize size, GCGLsizei inDepth, const PixelStoreParameters& parameters)
```
I'm not a fan of `inDepth`, but I can't think of a better name. Does WebKit have a 3D size type?

kkinnunen-apple replied:
> Does WebKit have a 3D size type?

No.

djg reviewed (approved):
:shipit:

kkinnunen-apple committed:
Row byte count is not available in GraphicsContextGL::computeImageSizeBytes
https://bugs.webkit.org/show_bug.cgi?id=257452
rdar://109969560

Reviewed by Dan Glastonbury.

GraphicsContextGL::computeImageSizeInBytes would return the image size
but not the count of row bytes. The row bytes are needed for GPUP side
implementation where the GPUP is queried for tightly packed image
and then the image data is expanded by WP to the client buffer.

Rename to computeImageSize.

Simplify the computation by using Checked all the way to the end.
The idea of Checked is that the intermediate values do not need to be
checked.

Return a struct instead of numerous output variables. These were
error-prone, as they were written even though the function would return
error.

Remove the unused paddingInBytes, which was not used.
Add row bytes variables.

Add checked variant of roundUpToMultipleOfNonPowerOfTwo.
Add unchecked variant, for consistency.
Fix overflow in bmalloc::roundUpToMultipleOfNonPowerOfTwo and
untemplatize, as the template parameter could not be anything else than
size_t.

* Source/WTF/wtf/StdLibExtras.h:
(WTF::roundUpToMultipleOfNonPowerOfTwo):
* Source/WebCore/html/canvas/WebGL2RenderingContext.cpp:
(WebCore::WebGL2RenderingContext::readPixels):
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::readPixels):
(WebCore::WebGLRenderingContextBase::validateImageFormatAndType):
(WebCore::WebGLRenderingContextBase::validateTexFuncData):
* Source/WebCore/html/canvas/WebGLRenderingContextBase.h:
* Source/WebCore/platform/graphics/GraphicsContextGL.cpp:
(WebCore::GraphicsContextGL::computeImageSize):
(WebCore::GraphicsContextGL::packImageData):
(WebCore::GraphicsContextGL::extractPixelBuffer):
(WebCore::GraphicsContextGL::extractTextureData):
(WebCore::GraphicsContextGL::computeImageSizeInBytes): Deleted.
* Source/WebCore/platform/graphics/GraphicsContextGL.h:
* Source/bmalloc/bmalloc/Algorithm.h:
(bmalloc::roundUpToMultipleOfNonPowerOfTwo):
* Tools/TestWebKitAPI/Tests/WTF/StdLibExtras.cpp:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/264794@main
